### PR TITLE
create gcloud activities

### PR DESCRIPTION
**Pull request by CJ-Wright**

### CODE REVIEW

**Comment by CJ-Wright** [comment]
@scopatz

File: rever/activities/gcloud.xsh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    def _func(self):
+        """Deploys the build docker containter to the google cloud"""
+        # make sure we are logged in
+        _ensure_default_credentials()
+        account = _ensure_account()
+        # get cluster credentials
+        ![gcloud container clusters get-credentials --account @(account) \
+          --zone=$GCLOUD_ZONE --project=$GCLOUD_PROJECT_ID $GCLOUD_CLUSTER]
+        # set new image
+        ![kubectl set image deployment/$GCLOUD_CONTAINER_NAME $GCLOUD_CONTAINER_NAME=docker.io/$GCLOUD_DOCKER_ORG/$GCLOUD_DOCKER_REPO:$VERSION]
```
Comment (CJ-Wright) [inline rever/activities/gcloud.xsh]: Can I get away with this? (An env var as a kwarg key)

File: rever/activities/gcloud.xsh [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        This activity may be configured with the following environment variables:
+
+        :$GCLOUD_PROJECT_ID: str, the gcloud project id
+        :$GCLOUD_ZONE: str, the gcloud zone
+        """
+
+    def __init__(self, *, deps=frozenset()):
+        super().__init__(name='deploy_to_gcloud_app', deps=deps, func=self._func,
+                         desc="Deploys an app to the google cloud via the app engine", )
+        self._re_cache = {}
```
Comment (scopatz) [inline rever/activities/gcloud.xsh]: What is this for?

File: rever/activities/gcloud.xsh [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        account = _ensure_account()
+        # get cluster credentials
+        ![gcloud container clusters get-credentials --account @(account) \
+          --zone=$GCLOUD_ZONE --project=$GCLOUD_PROJECT_ID $GCLOUD_CLUSTER]
+        # set new image
+        ![kubectl set image deployment/$GCLOUD_CONTAINER_NAME $GCLOUD_CONTAINER_NAME=docker.io/$GCLOUD_DOCKER_ORG/$GCLOUD_DOCKER_REPO:$VERSION]
+
+
+class DeployTOGCloudApp(Activity):
+    """Deploys an app to the google cloud via the app engine
```
Comment (scopatz) [inline rever/activities/gcloud.xsh]: This isn't google's app engine.  That is something different

File: rever/activities/gcloud.xsh (hunk 1 above)
Comment (scopatz) [reply rever/activities/gcloud.xsh]: This looks fine, but what is the keyword argument?

File: rever/activities/gcloud.xsh [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    This activity may be configured with the following environment variables:
+
+    :$GCLOUD_PROJECT_ID: str, the gcloud project id
+    :$GCLOUD_ZONE: str, the gcloud zone
+    :$GCLOUD_CLUSTER: str the kubernetes cluster to deploy to
+    :$GCLOUD_CONTAINER_NAME: str, the name of the container image to deploy to
+    :$GCLOUD_DOCKER_ORG: str, the name of the docker org to pull the container
+        from
+    :$GCLOUD_DOCKER_REPO: str, the name of the docker container repo to use
```
Comment (scopatz) [inline rever/activities/gcloud.xsh]: The above env vars should be specified in environ.py

File: rever/activities/gcloud.xsh (hunk 3 above)
Comment (scopatz) [reply rever/activities/gcloud.xsh]: Or sorry, maybe this is the app engine

File: rever/activities/gcloud.xsh (hunk 2 above)
Comment (CJ-Wright) [reply rever/activities/gcloud.xsh]: (copy paste error)

File: rever/activities/gcloud.xsh (hunk 1 above)
Comment (CJ-Wright) [reply rever/activities/gcloud.xsh]: this thing
`$GCLOUD_CONTAINER_NAME=docker.io/$GCLOUD_DOCKER_ORG/$GCLOUD_DOCKER_REPO:$VERSION`
So the kwarg is the location on docker of the image.

**Comment by scopatz** [comment]
Thanks for putting this in @CJ-Wright! I think that both of these need to have check methods and specify which CLI tools they use.  You can check out some of the other activities that use checks for examples

File: rever/activities/gcloud.xsh (hunk 1 above)
Comment (scopatz) [reply rever/activities/gcloud.xsh]: 1. All xonsh is going to do is expand that string.

   ```sh
   scopatz@artemis ~ $ $A = "aaa"
   scopatz@artemis ~ $ $B = "bbb"
   scopatz@artemis ~ $ echo $A=docker.io/$B
  aaa=docker.io/bbb
   ```

2. This isn't a keyword argument. There is no assignment happening. It is just an argument to a subprocess call.  Sorry, If I am being pedantic here, but I am trying to figure out if you what that value assigned to an env var or just want the env var expanded.

**Comment by CJ-Wright** [comment]
Is there anything in particular you want to "specify which CLI tools they use" beyond the check stuff.

File: rever/activities/gcloud.xsh [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        _ensure_default_credentials()
+        account = _ensure_account()
+        # get cluster credentials
+        ![gcloud container clusters get-credentials --account @(account) \
+          --zone=$GCLOUD_ZONE --project=$GCLOUD_PROJECT_ID $GCLOUD_CLUSTER]
+        # set new image
+        ![kubectl set image deployment/$GCLOUD_CONTAINER_NAME $GCLOUD_CONTAINER_NAME=docker.io/$GCLOUD_DOCKER_ORG/$GCLOUD_DOCKER_REPO:$VERSION]
+
+
+class DeployTOGCloudApp(Activity):
```
Comment (scopatz) [inline rever/activities/gcloud.xsh]: should `TO` be `To`?

File: rever/environ.xsh [hunk 6]
```diff
@@ -51,6 +51,7 @@ def default_dag():
     from rever.activities.tag import Tag
     from rever.activities.push_tag import PushTag
     from rever.activities.version_bump import VersionBump
+    from rever.activities.gcloud import DeployToGCloud, DeployTOGCloudApp
```
Comment (scopatz) [inline rever/environ.xsh]: TO -> to, and below

**Comment by scopatz** [comment]
> Is there anything in particular you want to "specify which CLI tools they use" beyond the check stuff.

Nope just the check stuff!  Also, is there a reasonable way we can test this?

**Comment by CJ-Wright** [comment]
Not sure about testing, maybe google will let us to hello world apps? (eg make a test/example which uses their tutorial + rever)

File: rever/activities/gcloud.xsh [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        bad = []
+        for cli, package in clis:
+            if not !(which @(cli)):
+                bad.append((cli, package))
+        if bad:
+            s = ''
+            for cli, package in bad:
+                s += f'Could not find {cli}! Try installing:\n  $ conda install {package}'
+            print(s)
+            raise RuntimeError(s)
```
Comment (scopatz) [inline rever/activities/gcloud.xsh]: You don't need to write this boiler plate code and instead should use the `requires` kwarg of the activity init method: https://regro.github.io/rever-docs/api/activity.html

File: rever/activities/gcloud.xsh (hunk 1 above)
Comment (CJ-Wright) [reply rever/activities/gcloud.xsh]: I just want the env var expanded

**Comment by scopatz** [comment]
Yeah, if there is an anonymous hello world that doesn't need credentials, that's be great! Otherwise, maybe you can alias `gcloud` and `kubctrl` to `echo`, like we do in some of the other tests.

File: rever/activities/gcloud.xsh (hunk 1 above)
Comment (scopatz) [reply rever/activities/gcloud.xsh]: Then yes, you are good!

**Comment by CJ-Wright** [comment]
do you have an example of ``echo`` in the tests?

**Comment by scopatz** [comment]
Looks like it is a fixture, defined here: https://github.com/regro/rever/blob/master/tests/conftest.py#L37

used here: https://github.com/regro/rever/blob/master/tests/test_conda_forge_activity.py#L242

**Comment by CJ-Wright** [comment]
@scopatz any idea why there is a `KeyError: 'Unknown environment variable: $LOGGER'` for the gcloud tests?

**Comment by scopatz** [comment]
You are not using the `gitrepo` fixture to set up a proper context

**Comment by scopatz** [comment]
LGTM, anything else you want to change?

**Comment by CJ-Wright** [comment]
No I think this is good (we tested the app one today and it seemed to work properly). Do you want a squash? (I guess we should run a release and then change metachannel to use the new activity)

**Comment by scopatz** [comment]
Squash is evil :)

**Comment by scopatz** [comment]
Thanks @CJ-Wright!

**Comment by CJ-Wright** [comment]
Thank you!

**Pull request merged**